cpp/src/arrow/compute/kernels/mask-test.cc

```diff
@@ -0,0 +1,145 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// returnGegarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <vector>
+
+#include "arrow/compute/context.h"
+#include "arrow/compute/kernels/mask.h"
+#include "arrow/compute/test-util.h"
+#include "arrow/testing/gtest_common.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/random.h"
+#include "arrow/testing/util.h"
+
+namespace arrow {
+namespace compute {
+
+using util::string_view;
+
+template <typename ArrowType>
+class TestMaskKernel : public ComputeFixture, public TestBase {
+ protected:
+  void AssertMaskArrays(const std::shared_ptr<Array>& values,
+                        const std::shared_ptr<Array>& mask, MaskOptions options,
+                        const std::shared_ptr<Array>& expected) {
+    std::shared_ptr<Array> actual;
+    ASSERT_OK(arrow::compute::Mask(&this->ctx_, *values, *mask, options, &actual));
+    AssertArraysEqual(*expected, *actual);
+  }
+  void AssertMask(const std::shared_ptr<DataType>& type, const std::string& values,
+                  const std::string& mask, MaskOptions options,
+                  const std::string& expected) {
+    std::shared_ptr<Array> actual;
+    ASSERT_OK(this->Mask(type, values, mask, options, &actual));
+    AssertArraysEqual(*ArrayFromJSON(type, expected), *actual);
+  }
+  Status Mask(const std::shared_ptr<DataType>& type, const std::string& values,
+              const std::string& mask, MaskOptions options, std::shared_ptr<Array>* out) {
+    return arrow::compute::Mask(&this->ctx_, *ArrayFromJSON(type, values),
+                                *ArrayFromJSON(boolean(), mask), options, out);
+  }
+};
+
+class TestMaskKernelWithNull : public TestMaskKernel<NullType> {
+ protected:
+  void AssertMask(const std::string& values, const std::string& mask, MaskOptions options,
+                  const std::string& expected) {
+    TestMaskKernel<NullType>::AssertMask(utf8(), values, mask, options, expected);
+  }
+};
+
+TEST_F(TestMaskKernelWithNull, MaskNull) {
+  MaskOptions options;
+  this->AssertMask("[null, null, null]", "[0, 1, 0]", options, "[null]");
+  this->AssertMask("[null, null, null]", "[1, 1, 0]", options, "[null, null]");
```

Comment:
Also should add tests with an empty mask array.

Reply:
will do